scripts/generate_matrix.py

```diff
@@ -0,0 +1,227 @@
+# !/usr/bin/env python3
+"""
+Script to generate a test matrix of Python and package versions for CI testing.
+
+This script fetches the list of supported Python versions and the versions of a
+specified package available on PyPI, determines compatible combinations based
+on package 'requires_python' metadata, and outputs a JSON matrix suitable for
+use in CI workflows.
+
+Features:
+- Generalized to work with any package (not just NumPy).
+- Caches HTTP requests to reduce network calls.
+- Prevents code duplication by using reusable functions.
+- Includes thorough doctests and documentation.
+
+Usage:
+    python generate_matrix.py [PACKAGE_NAME]
+
+Example:
+    python generate_matrix.py numpy
+
+"""
+
+import dataclasses
+import importlib.metadata
+import json
+import sys
+import urllib.error
+import urllib.request
+from functools import lru_cache
+
+# Constants for caching
+CACHE_SIZE = 128  # Adjust as needed
+
+
+@dataclasses.dataclass(frozen=True, eq=True, order=True)
+class Version:
+    major: int
+    minor: int
+    micro: int
+
+    def __str__(self) -> str:
+        return f"{self.major}.{self.minor}.{self.micro}"
+
+    @property
+    def minor_version(self) -> tuple[int, int]:
+        return self.major, self.minor
+
+    @classmethod
+    def from_str(cls, version_str: str) -> "Version":
+        if version_str.count(".") == 1:
+            version_str += ".0"
+        return cls(*map(int, version_str.strip().split(".")))
+
+
+def get_minimum_python_version(package: str) -> Version:
+    """
+    Fetch the minimum Python version specified in pyproject.toml.
+
+    Returns:
+        str: The minimum required version of the package.
+
+    Examples:
+        >>> get_minimum_python_version()  # doctest: +ELLIPSIS
+        Version(major=3, ...)
+    """
+    raw_version = importlib.metadata.metadata(package)["Requires-Python"]
+    if "<" in raw_version:
+        raise NotImplementedError("Version specifier with upper bound not yet supported!")
+
+    return Version.from_str(raw_version.replace(">=", "").replace("~=", ""))
+
+
+def get_minimum_version(package: str, dependency: str) -> Version:
+    """
+    Fetch the minimum dependency version required by the given package.
+
+    Args:
+        package (str): The package name ('scipy' or package name).
+        dependency (str): The dependency name ('numpy' or dependency name).
+
+    Returns:
+        str: The minimum required version of the package.
+
+    Examples:
+        >>> get_minimum_version("scipy", "numpy")  # doctest: +ELLIPSIS
+        Version(major=1, ...)
+    """
+    np_minimum_dep = next(
+        req for req in importlib.metadata.requires(package) if req.startswith(dependency) and " extra " not in req
+    )
+    np_minimum_dep = next(ver for ver in np_minimum_dep.split(",") if ">" in ver)
+    return Version.from_str(np_minimum_dep.replace(dependency, "").replace(">=", "").replace(">", ""))
```

Comment:
And this looks a lot like `scipy_minimum_numpy()` in `scripts/version_bounds.py`, like the one above